src/prng.rs

```diff
@@ -7,51 +7,109 @@
 
-pub(crate) const SEED_LENGTH: usize = 2 * BLOCK_SIZE;
-
-pub(crate) fn secret_share<F: FieldElement>(share1: &mut [F]) -> Result<Vec<u8>, getrandom::Error> {
-    let mut seed = vec![0; SEED_LENGTH];
-    getrandom(seed.as_mut_slice())?;
+pub(crate) fn secret_share<F: FieldElement>(share1: &mut [F]) -> Result<Vec<u8>, PrngError> {
+    let mut seed = [0; Aes128Ctr::STREAM_CIPHER_SEED_SIZE];
+    getrandom(&mut seed)?;
 
     // get prng array
-    let data: Vec<F> = Prng::new_with_seed_and_optional_length(&seed, Some(share1.len())).collect();
+    let data: Vec<F> = Prng::<_, Aes128Ctr>::new_with_seed(&seed)?
+        .take(share1.len())
+        .collect();
 
     // secret share
     for (s1, d) in share1.iter_mut().zip(data.iter()) {
         *s1 -= *d;
     }
 
-    Ok(seed)
+    Ok(seed.to_vec())
 }
 
 pub(crate) fn extract_share_from_seed<F: FieldElement>(
     length: usize,
     seed: &[u8],
 ) -> Result<Vec<F>, PrngError> {
-    if seed.len() != 2 * BLOCK_SIZE {
-        return Err(PrngError::SeedLen);
+    Ok(Prng::<_, Aes128Ctr>::new_with_seed(seed)?
+        .take(length)
+        .collect())
+}
+
+/// Generate a vector of uniform random field elements.
+pub fn random_vector<F: FieldElement>(len: usize) -> Result<Vec<F>, PrngError> {
+    Ok(Prng::<_, Aes128Ctr>::new()?.take(len).collect())
+}
+
+/// A stream cipher maps a short key to a stream of pseudorandom bytes.
+pub(crate) trait StreamCipher: Sized {
+    const STREAM_CIPHER_SEED_SIZE: usize;
+
+    /// Returns a new stream ciphered keyed by `key`.
+    fn new(key: &[u8]) -> Result<Self, PrngError>;
+
+    /// Fills the input buffer `buf` with the next `buf.len()` bytes of the stream.
+    fn fill(&mut self, buf: &mut [u8]);
+}
+
+impl StreamCipher for Aes128Ctr {
+    const STREAM_CIPHER_SEED_SIZE: usize = 32;
+
+    fn new(key: &[u8]) -> Result<Self, PrngError> {
+        if key.len() != Self::STREAM_CIPHER_SEED_SIZE {
+            return Err(PrngError::SeedSize);
+        }
+
+        // The first 16 bytes of the key and the last 16 bytes of the key are used, respectively,
+        // for the key and initialization vector for AES128 in CTR mode.
+        let aes_ctr_key = GenericArray::from_slice(&key[..aes::BLOCK_SIZE]);
+        let aes_ctr_iv = GenericArray::from_slice(&key[aes::BLOCK_SIZE..]);
+        Ok(Self::from_block_cipher(
+            Aes128::new(aes_ctr_key),
+            aes_ctr_iv,
+        ))
     }
 
-    Ok(Prng::new_with_seed_and_optional_length(seed, Some(length)).collect())
+    fn fill(&mut self, buf: &mut [u8]) {
+        self.apply_keystream(buf);
+    }
+}
+
+impl StreamCipher for blake3::OutputReader {
+    const STREAM_CIPHER_SEED_SIZE: usize = 32;
+
+    fn new(key: &[u8]) -> Result<Self, PrngError> {
+        if key.len() != Self::STREAM_CIPHER_SEED_SIZE {
+            return Err(PrngError::SeedSize);
+        }
+
+        let mut key_copy = [0; Self::STREAM_CIPHER_SEED_SIZE];
+        key_copy.clone_from_slice(key);
```

Comment:
Is this copy needed just to go from `&[u8]` to `&[u8; 32]`? I think you could skip this if you had `StreamCipher::new(key: &[u8; Self::STREAM_CIPHER_SEED_SIZE]`.

Reply:
Sadly this doesn't work (see comment above)